storage/dataflux/fast_list.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dataflux
+
+import (
+	"context"
+
+	"cloud.google.com/go/storage"
+)
+
+// listingMethod represents the method of listing.
+type listingMethod int
+
+const (
+	// open when any method can be used to list.
+	open listingMethod = iota
+	// sequential when the listing is done sequentially.
+	sequential
+	// worksteal when the listing is done using work stealing algorithm.
+	worksteal
+)
+
+// Input contains options for listing objects.
+type Input struct {
+	// Client is the storage client to list objects from.
+	Client *storage.Client
+	// BucketName is the name of the bucket to list objects from.
+	BucketName string
+	// Parallelism is number of parallel workers to use for listing.
+	Parallelism int
+	// pageSize is the number of objects to list.
+	BatchSize int
+	// query is the query to filter objects for listing.
+	Query storage.Query
+	// SkipDirectoryObjects is to indicate whether to list directory objects.
+	SkipDirectoryObjects bool
+}
+
+// Lister is used for interacting with Dataflux fast-listing.
+// The caller should initialize it with NewLister() instead of creating it directly.
+type Lister struct {
+	// method indicates the listing method(open, sequential, worksteal) to be used for listing.
+	method listingMethod
+	// pageToken is the token to use for sequential listing.
+	pageToken string
+	// ranges is the channel to store the start and end ranges to be listed by the workers in worksteal listing.
+	ranges chan *listRange
+	// bucket is the bucket handle to list objects from.
+	bucket *storage.BucketHandle
+	// parallelism is number of parallel workers to use for listing.
+	parallelism int
+	// batchSize is the number of objects to list.
+	batchSize int
+	// query is the query to filter objects for listing.
+	query storage.Query
+	// skipDirectoryObjects is to indicate whether to list directory objects.
+	skipDirectoryObjects bool
+}
+
+// CloseFunc is the function to close the range channel of a Lister.
+type CloseFunc func()
+
+// NewLister creates a new dataflux Lister to list objects in the give bucket.
+func NewLister(opts *Input) (*Lister, CloseFunc) {
+
+	lister := &Lister{}
+	return lister, func() { lister.Close() }
+}
+
+// NextBatch runs worksteal algorithm and sequential listing in parallel to quickly
```

Comment:
I think this isn't completely clear on how one would call this. Maybe it's helpful to add an `example_test.go` file like we have in storage/ and transfermanager in order to make an example for the docs?

Reply:
Also is there a way for the user to choose which algorithm is used, or do you always run both?

Reply:
Added example_test.go. 


Users cannot choose the algorithm. When NextBatch is called for the first time by a lister, both the algorithm runs. Whichever algorithm completes first will be used for subsequent runs.